scripts/debug_tools/reduce_ctu_crash.py

```diff
@@ -0,0 +1,625 @@
+                       analyzer command, and the clang binary; etc.
+
+    Returns a list of commands that will be written to a bash script file.
+    Each command is a list itself, divided along spaces.
+    """
+    conditions = []
+    ending = ['2>&1', '&&\\\n']
+    is_dependent = (reduced_file != opts.analyzed_file)
+
+    # Ensure that the file is compilable, then clean up.
+    object_file = os.path.splitext(reduced_file)[0] + '.o'
+    conditions.extend([opts.clang, '-c', std_flag, reduced_file,
+                       '-o', object_file, '>/dev/null'] + ending)
+    conditions.extend(['rm', object_file, '>/dev/null'] + ending)
+
+    if is_dependent:
+        # Re-generate AST in "workspace"/report-debug/ctu-dir.
+        triple_arch = prepare_cmds.get_triple_arch(opts.analyze_sh)
+        conditions.extend(get_ast_dump_cmd(opts.clang, std_flag,
+                                           reduced_file, triple_arch))
+        conditions.extend(['>/dev/null'] + ending)
+
+    # Add negation of the failing analyzer command.
+    failing_cmd = get_failing_cmd(opts.analyze_sh, opts.analyzed_file)
+    err_dest = os.path.join(os.path.dirname(reduced_file), 'crash.txt')
+    conditions.extend(['!'] + failing_cmd + ['>' + err_dest] + ending)
+
+    # Ensure that the previous command still produces the assertion.
+    assert_string = opts.assert_string.replace('\"', '\\\"') \
+                                      .replace('`', '\\`')   \
+                                      .replace('!', '\\!')
+    conditions.extend(['grep', '\"' + assert_string + '\"',
+                       err_dest, '>/dev/null', '2>&1'])
+
+    return dump_test(conditions, reduced_file)
+
+
+def reduce_iteration(opts, cmds, counter, work_dir):
+    """
+    Go through all source files in the ZIP, running one round of C-Reduce
+    on them. Start with dependent files. Before running C-Reduce, check
+    if an empty version of a file satisfies the interestingness test.
+    """
+    LOG.info('Starting to reduce dependent files.')
+    for cmd in cmds:
+        if not is_cpp_or_c_file(cmd['file']) or \
+            cmd['file'] == opts.analyzed_file:
+            continue
+        std_flag = get_std_flag(cmd['command'])
+        if os.stat(cmd['file']).st_size > 0:
+            LOG.info('Reducing dependent file: %s', cmd['file'])
+            test = assemble_test(cmd['file'], opts, std_flag)
+            run_creduce(cmd['file'], test, opts.jobs, counter)
+            run_ctu_collect(opts.compilation_db, opts.jobs, work_dir)
+
+    LOG.info('Starting to reduce analyzed file: %s', opts.analyzed_file)
+    test = assemble_test(opts.analyzed_file, opts, opts.analyze_std)
+    run_creduce(opts.analyzed_file, test, opts.jobs, counter)
+
+
+def get_cmd_after_preproc(old_cmd, file_dir_path):
+    """
+    Create a new compilation command for a preprocessed and relocated
+    source file.
+    """
+    old_cmd = __remove_dependency_generation(old_cmd.split(" "))
```

Comment:
using [shlex.split](https://docs.python.org/3/library/shlex.html) is generally a better idea to split up compilation commands there might be arguments with space in them like `-Dvar="my value"` which would be split wrong..